File: lib/src/main/java/me/ameriod/lib/mvp/view/MvpFragment.java
package me.ameriod.lib.mvp.view;

import android.app.Fragment;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import me.ameriod.lib.mvp.Mvp;
import me.ameriod.lib.mvp.view.deligate.FragmentDelegate;
import me.ameriod.lib.mvp.view.deligate.FragmentDelegateCallback;
import me.ameriod.lib.mvp.view.deligate.FragmentDelegateImpl;

public abstract class MvpFragment<V extends Mvp.View, P extends Mvp.Presenter<V>> extends Fragment
        implements Mvp.View, FragmentDelegateCallback<V, P> {

    private FragmentDelegate<V, P> fragmentDelegate;

    @Override
    @SuppressWarnings("unchecked")
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (fragmentDelegate == null) {
            fragmentDelegate = new FragmentDelegateImpl<>(this);
        }
        fragmentDelegate.onViewCreated(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        fragmentDelegate.onSaveInstanceState(outState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        fragmentDelegate.onDestroyView();
    }

    @NonNull
    @Override
    public abstract P createPresenter();

    @Override
    public P getPresenter() {
        return fragmentDelegate.getPresenter();
    }

    @Override
    @SuppressWarnings("unchecked")
    public V getMvpView() {
        return (V) this;
    }
}

File: lib/src/main/java/me/ameriod/lib/mvp/view/MvpFragment.kt
package me.ameriod.lib.mvp.view

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.View
import me.ameriod.lib.error.Error
import me.ameriod.lib.error.ErrorDisplayDelegate
import me.ameriod.lib.error.ErrorDisplayDelegateImpl

import me.ameriod.lib.mvp.Mvp
import me.ameriod.lib.mvp.deligate.FragmentDelegate
import me.ameriod.lib.mvp.deligate.FragmentDelegateCallback
import me.ameriod.lib.mvp.deligate.FragmentDelegateImpl

abstract class MvpFragment<V : Mvp.View, out P : Mvp.Presenter<V>> : androidx.fragment.app.Fragment(), Mvp.View,
        FragmentDelegateCallback<V, P> {

    private var fragmentDelegate: FragmentDelegate<V, P>? = null
    private var errorDisplayDelegate : ErrorDisplayDelegate? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (fragmentDelegate == null) {
            fragmentDelegate = FragmentDelegateImpl(this)
        }
        fragmentDelegate!!.onViewCreated(savedInstanceState)

        if (errorDisplayDelegate == null) {
            errorDisplayDelegate = ErrorDisplayDelegateImpl()
        }
        errorDisplayDelegate!!.attachView(view)

    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        fragmentDelegate!!.onSaveInstanceState(outState)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        fragmentDelegate!!.onDestroyView()
    }

    abstract override fun createPresenter(): P

    override fun getPresenter(): P {
        return fragmentDelegate!!.getPresenter()
    }

    @Suppress("UNCHECKED_CAST")
    override fun getMvpView(): V {
        return this as V
    }

    fun getErrorDisplayDelegate() : ErrorDisplayDelegate = errorDisplayDelegate!!

    override fun displayErrorMessage(error: Error<*>) {
        errorDisplayDelegate!!.displayError(error)
    }

    override fun displayError(error: String) {
        displayErrorMessage(Error.SnackbarMessage(error))
    }
}
